### DISCUSSION TITLE
Can the color of an existing svg be changed from code?

### DISCUSSION THREAD

**Discussion by pauldendulk** [wieslawsoltes/Svg.Skia#166] [Q&A] [answered]
Sorry for being lazy and not properly reading the documentation. For our project it would be really helpful if we could alter the color of an existing svg. So, for instance the fill style of a path like this: `style="fill:#cc0000"`.

If not how could such functionality be added? Where in the code should I look and would you be interested in a MR?

**Comment by wieslawsoltes** [comment, marked as answer]
I think you can do some manipulation but only after loading SvgDocument but but when it's rendering

**Comment by pauldendulk** [comment]
Thanks! We are not using the SvgDocument in our code so I need to did deeper into this library. I just cloned it on Windows but get by a lot of build errors if I open the sln in Visual Studio. Is there anything specific I need to do? Call some build command? The build.cmd and build.ps1 scripts fails complaining that Svg.Generators.csproj was not found. Should I generate some source somehow?

**Reply by wieslawsoltes** [reply]
You need to checkout submodules too

**Reply by pauldendulk** [reply]
Thanks! fixed it.

**Comment by pauldendulk** [comment]
I ended up using the class below. In our code we used `SKSvg.FromSvg(svg)` before. We need the SvgDocument that is used inside SKSvg for the manipulation. So, we had to do some of the things done inside SKSvg ourselves. 

The purpose of this code is to override the existing fill and stroke. Please tell me if there are more efficient ways of doing this, or more reuse of existing methods. For instance iterating over all SvgElements in a SvgDocument, or a simpler way to convert the SvgDocument to a SKPicture.

```csharp
using Svg.Model;
using Svg.Skia;
using Svg;
using SkiaSharp;
using System.Drawing;

namespace MN.MapsuiBuilder.LayerBuilders
{
    public class SvgLoader
    {
        public static SKPicture? ToSKPicture(Stream stream, Color? fillColor = null, Color? strokeColor = null)
        {
            var svgDocument = SvgExtensions.Open(stream);
            if (svgDocument is null) return null;

            var elements = GetAllElements(svgDocument.Children);
            foreach (var element in elements)
            {
                if (element.Fill is not null && fillColor is { })
                    element.Fill = new SvgColourServer(fillColor.Value);
                if (element.Stroke is not null && strokeColor is { })
                    element.Stroke = new SvgColourServer(strokeColor.Value);
            }

            return ToSKPicture(svgDocument);
        }

        private static SKPicture? ToSKPicture(SvgDocument svgDocument)
        {
            var skiaModel = new SkiaModel(new SKSvgSettings());
            var assetLoader = new SkiaAssetLoader(skiaModel);
            var model = SvgExtensions.ToModel(svgDocument, assetLoader, out var _, out _);
            return skiaModel.ToSKPicture(model);
        }

        public static List<SvgElement> GetAllElements(SvgElementCollection elements)
        {
            var result = new List<SvgElement>();
            foreach (var element in elements)
            {
                result.Add(element);

                if (element.Children.Count > 0)
                    result.AddRange(GetAllElements(element.Children));
            }
            return result;
        }
    }
}
```

**Comment by Drostvorn** [comment]
Hi pauldendulk, thanks for the solution to this. I was looking for the same thing.

I'm using this Avalonia.Svg.Skia lib to read files into my app, either of two ways:

	<Image Name="MenuExpandButtonSvg" Source="{SvgImage /Assets/hamburger_icon.svg}"/>
                                                OR
	<Svg Name="MenuExpandButtonSvg" Height="20" Width="20" Path="/Assets/hamburger_icon_dark.svg"/>

I'm wondering how I would turn either the Avalonia.Svg.Skia.Svg or Avalonia.Controls.Image into a Stream, to then call the SKToPicture(Stream stream, Color? fillColor = null, Color? strokeColor = null) method you wrote?

Thanks in advance for any info!

Also thanks much to wieslawsoltes! Nice library!

James

**Reply by thien-an-ly** [reply]
Hi Drostvorn,

I'm stumbling upon the same use case. Did you ever find out a workaround?

Currently I'm doing
```
var path = (iconResource as SvgSource)!.Path;
path = $"{Directory.GetCurrentDirectory()}{path}";
```
with `path` being the relative URI that works at design time, and calling the other method like so:
```
var svg = SvgCustomization.Apply(path!, color, color);
Icon = SvgSource.LoadFromSvgDocument(svg)!;

// ...
public static SvgDocument Apply(string path, Color? fillColor = null, Color? strokeColor = null)
{
    var svg = SvgExtensions.Open(path) ?? throw new IOException($"Unable to read from {path}");
    //...
}
```
unfortunately this approach doesn't work at runtime, because apparently Avalonia doesn't copy the static assets over to the build location. Other than that the overridden fill and stroke color seems to be working well.

Thanks in advance!
Aly

**Reply by pauldendulk** [reply]
Reading from file can work on some platforms, if you use the proper directory and set the file to `copy if newer`, but even that can be problematic on some platforms because of access rights. What I often do is to make the file an embedded resource and load the stream from the embedded resource. There are never problems with access rights. The nasty thing is that you need to get the path to the embedded resource correct. Here are some Mapsui helper functions that might be useful: https://github.com/Mapsui/Mapsui/blob/main/Mapsui/Utilities/EmbeddedResourceLoader.cs

**Reply by thien-an-ly** [reply]
Thanks, that's super helpful! I actually followed the embedded resource route in the first place, just never realized you could programmatically obtain a stream from it for other use. That's another toy in my toolbox 👍

In the end I decided to just use [this tool](https://github.com/narendragoudelli/SvgToPathGeometry) to convert all my SVGs into `<GeometryStream>` beforehand, embed them in an `Icons.axaml` style resource dictionary with shorthand keys, and `{StaticResource $key}` to them in a `<PathIcon.Data>`. The tool author said there might be some complex SVGs that can't be converted correctly, but for my use case so far so good.

`<PathIcon>` actually allows you to set `Foreground` to a specific color, which in my case is more than sufficient (originally just wanted some icons to pop out with the system default accent, heh). As far as I can tell, that seems to set both the `FillColor` and `StrokeColor`, so if you need the same degree of customization that preserving the original SVG format offers, maybe that's a dealbreaker?
